openmmtools/cache.py

```diff
@@ -387,8 +388,10 @@ def get_context(self, thermodynamic_state, integrator=None):
                 # Only one match.
                 context = self._lru[matching_context_ids[0]]
             else:
-                # Multiple matches, prefer non-default Integrator.
-                for context_id in matching_context_ids:
+                # Multiple matches, prefer the non-default Integrator.
+                # Always pick the least recently used to make two consective
+                # calls retrieving the same integrator.
+                for context_id in reversed(matching_context_ids):
                     if context_id[1] != self._default_integrator_id():
                         context = self._lru[context_id]
                         break
```

Comment:
Exactly.

> but the default can just be cloned and fed into multiple contexts

If it happens, it means that the thermodynamic states of the multiple contexts are incompatible, so only one will end up in `matching_context_ids`.